### PR TITLE
`data` `azurerm_shared_image_version` - consider `excludeFromLatest` for `latest` version

### LINKED ISSUE #10633: azurerm provider does not respect exclude_from_latest flag when creating a virtual machine using a shared image
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.6
* provider registry.terraform.io/hashicorp/azurerm v2.47.0

### Affected Resource(s)

* `azurerm_linux_virtual_machine`

### Terraform Configuration Files

```hcl
# Azure RM provider
provider "azurerm" {
  features {}
  alias           = "alias"
  tenant_id       = "secret"
  subscription_id = "secret"
}

# Shared image version
data "azurerm_shared_image_version" "image" {
  name                = "latest"
  image_name          = "ubuntu2004"
  gallery_name        = "test_imagegallery"
  resource_group_name = "test_rg"
  provider            = azurerm.alias
}

# Virtual machine to create based on shared image version
resource "azurerm_linux_virtual_machine" "test_vm" {
  name                  = "test_vm"
  resource_group_name   = "test_rg"
  location              = "WestEurope"
  network_interface_ids = [azurerm_network_interface.test_vm_nic.id]
  size                  = "Standard_DS1_v2"

  os_disk {
    name              = "test_vm_disk"
    caching           = "ReadWrite"
    storage_account_type = "Premium_LRS"
  }

  source_image_id    = data.azurerm_shared_image_version.image.id

  computer_name  = "test_vm"

  admin_username = "test"
  admin_ssh_key {
    username = "test"
    public_key = "ssh-rsa ~snip~"
  }
}

resource "azurerm_network_interface" "test_vm_nic" {
  name                        = "test_vm_nic"
  location                    = "WestEurope"
  resource_group_name         = "test_rg"

  ip_configuration {
    name                          = "test_vm_nic_config"
    subnet_id                     = ~snip~
    private_ip_address_allocation = "Dynamic"
  }
}
```

```hcl
# packer_build_image.hcl
source "azure-arm" "ubuntu_20-04" {
  image_offer                       = "0001-com-ubuntu-server-focal"
  image_publisher                   = "Canonical"
  image_sku                         = "20_04-lts-gen2"
  location                          = "WestEurope"
  managed_image_name                = "ubuntu2004"
  managed_image_resource_group_name = "test_imagegallery_rg"
  os_type                           = "Linux"
  subscription_id                   = "secret"
  tenant_id                         = "secret"
  vm_size                           = "Standard_DS1_v2"
  
  shared_gallery_image_version_exclude_from_latest = true

  shared_image_gallery_destination {
    image_name          = "ubuntu2004"
    image_version       = var.image_version # increment of 0.0.x with every packer build
    gallery_name        = "test_imagegallery"
    resource_group      = "test_imagegallery_rg"
  }
}

build {
  sources = ["source.azure-arm.ubuntu_20-04"]
}
```

### Debug Output

Output from packer build, baking a 0.0.6 version that should be excluded from latest
```
azure-arm.ubuntu_20-04: Publishing to Shared Image Gallery ...
azure-arm.ubuntu_20-04:  -> MDI ID used for SIG publish           : '~snip~/images/ubuntu2004'
azure-arm.ubuntu_20-04:  -> SIG publish resource group            : 'test_rg'
azure-arm.ubuntu_20-04:  -> SIG gallery name                      : 'test_imagegallery'
azure-arm.ubuntu_20-04:  -> SIG image name                        : 'ubuntu2004'
azure-arm.ubuntu_20-04:  -> SIG image version                     : '0.0.6'
azure-arm.ubuntu_20-04:  -> SIG replication regions               : '[West Europe]'
azure-arm.ubuntu_20-04:  -> SIG image version endoflife date      : ''
azure-arm.ubuntu_20-04:  -> SIG image version exclude from latest : 'true'
azure-arm.ubuntu_20-04:  -> SIG replica count [1, 10]             : '1'
azure-arm.ubuntu_20-04:  -> Shared Gallery Image Version ID       : '~snip~/images/ubuntu2004/versions/0.0.6'
```

```
Terraform will perform the following actions:

  # azurerm_linux_virtual_machine.test_vm must be replaced
-/+ resource "azurerm_linux_virtual_machine" "test_vm" {
      ~ id                              = "~snip~/test_vm" -> (known after apply)
        name                            = "test_vm"
      ~ private_ip_address              = "10.40.1.4" -> (known after apply)
      ~ virtual_machine_id              = "~snip~" -> (known after apply)
      ~ source_image_id                 = "~snip~/images/ubuntu2004/versions/0.0.6" # forces replacement
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

### Expected Behaviour

Terraform creates a virtual machine based on shared image version 0.0.4

### Actual Behaviour

Terraform created a virtual machine based on version 0.0.6

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

We have an Azure image gallery, which contains a definition with multiple versions of an image.
* 0.0.1
* 0.0.2
* 0.0.3
* 0.0.4
* 0.0.5 (exclude from latest flag enabled)
* 0.0.6 (exclude from latest flag enabled)

Using the portal to create a virtual machine using the same gallery, gives me a virtual machine based on version 0.0.4

The bug seems similar to these bugs but considering I'm using terraform v0.14.6 + azurerm v2.47.0, bug is either different or still present.
* https://github.com/terraform-providers/terraform-provider-azurerm/issues/10516
* https://github.com/terraform-providers/terraform-provider-azurerm/issues/10374

### PR DESCRIPTION
**Pull request by myc2h6o**
Fix #10633

When a shared image version is created with `excludeFromLatest` set to `false`, on Azure Portal or when creating a VM, this version is not considered. Data source `azurerm_shared_image_version` ignores this aspect, which seems not aligned with the behavior. Adding the logic to find the most latest version which is not marked as `excludeFromLatest`. Also updating the test cases so that ordering is also tested, old test cases didn't test the ordering.

### CODE REVIEW

**Comment by myc2h6o** [comment]
Test result
image

**Review by jackofallops — DISMISSED** [review]
Hi @myc2h6o - Thanks for this. I think we need to feature flag this change as the output will be behaviourally different so users may not get their expected results.

File: internal/services/compute/shared_image_version_data_source.go [hunk 1]
```diff
@@ -193,8 +193,11 @@ func obtainImage(client *compute.GalleryImageVersionsClient, ctx context.Context
 					return nil, fmt.Errorf("parsing version(s): %v", errs)
 				}
 			}
-			image := images[len(images)-1]
-			return &image, nil
+			for i := len(images) - 1; i >= 0; i-- {
+				if prop := images[i].GalleryImageVersionProperties; prop == nil || prop.PublishingProfile == nil || prop.PublishingProfile.ExcludeFromLatest == nil || !*prop.PublishingProfile.ExcludeFromLatest {
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply internal/services/compute/shared_image_version_data_source.go:204]: yes make sense, I've moved this behind the 4.0 flag

**Comment by myc2h6o** [comment]
Updated test result
image

File: website/docs/d/shared_image_version.html.markdown [hunk 2]
```diff
@@ -30,6 +30,8 @@ The following arguments are supported:
 
 ~> **Note:** You may specify `latest` to obtain the latest version or `recent` to obtain the most recently updated version.
 
+~> **Note:** In 3.0, `latest` may return the image version with `exclude_from_latest` set to `true`. Starting from 4.0, `latest` does not return the image version with `exlude_from_latest` set to `true`.
```
Comment (katbyte) [inline website/docs/d/shared_image_version.html.markdown:33]:
```suggestion
~> **Note:** In 3.0, `latest` may return an image version with `exclude_from_latest` set to `true`. Starting from 4.0 onwards `latest` will not return image versions with `exlude_from_latest` set to `true`.
```

**Review by katbyte — APPROVED** [review]
aside from 1 omment LGTM 💾

File: internal/services/compute/shared_image_version_data_source.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ASHR4) [reply internal/services/compute/shared_image_version_data_source.go:204]: @myc2h6o When will 4.0 be released, so this provider has this functionality?

**Pull request merged**